src/ray/raylet/raylet.cc

```diff
@@ -45,63 +45,57 @@ Raylet::Raylet(boost::asio::io_service &main_service, const std::string &socket_
+      object_manager_(main_service, self_node_id_, object_manager_config,
+                      object_directory_),
+      node_manager_(main_service, self_node_id_, node_manager_config, object_manager_,
+                    gcs_client_, object_directory_),
       socket_name_(socket_name),
       acceptor_(main_service, boost::asio::local::stream_protocol::endpoint(socket_name)),
       socket_(main_service) {
-  // Start listening for clients.
-  DoAccept();
-
-  RAY_CHECK_OK(RegisterGcs(
-      node_ip_address, socket_name_, object_manager_config.store_socket_name,
-      redis_address, redis_port, redis_password, main_service, node_manager_config));
-
-  RAY_CHECK_OK(RegisterPeriodicTimer(main_service));
+  self_node_info_.set_node_id(self_node_id_.Binary());
+  self_node_info_.set_node_manager_address(node_ip_address);
+  self_node_info_.set_raylet_socket_name(socket_name);
+  self_node_info_.set_object_store_socket_name(object_manager_config.store_socket_name);
+  self_node_info_.set_object_manager_port(object_manager_.GetServerPort());
+  self_node_info_.set_node_manager_port(node_manager_.GetServerPort());
 }
 
 Raylet::~Raylet() {}
 
-ray::Status Raylet::RegisterPeriodicTimer(boost::asio::io_service &io_service) {
-  boost::posix_time::milliseconds timer_period_ms(100);
-  boost::asio::deadline_timer timer(io_service, timer_period_ms);
-  return ray::Status::OK();
+void Raylet::Start() {
+  RAY_CHECK_OK(RegisterGcs());
+
+  // Start listening for clients.
+  DoAccept();
 }
 
-ray::Status Raylet::RegisterGcs(const std::string &node_ip_address,
-                                const std::string &raylet_socket_name,
-                                const std::string &object_store_socket_name,
-                                const std::string &redis_address, int redis_port,
-                                const std::string &redis_password,
-                                boost::asio::io_service &io_service,
-                                const NodeManagerConfig &node_manager_config) {
-  GcsNodeInfo node_info = gcs_client_->client_table().GetLocalClient();
-  node_info.set_node_manager_address(node_ip_address);
-  node_info.set_raylet_socket_name(raylet_socket_name);
-  node_info.set_object_store_socket_name(object_store_socket_name);
-  node_info.set_object_manager_port(object_manager_.GetServerPort());
-  node_info.set_node_manager_port(node_manager_.GetServerPort());
-
-  RAY_LOG(DEBUG) << "Node manager " << gcs_client_->client_table().GetLocalClientId()
-                 << " started on " << node_info.node_manager_address() << ":"
-                 << node_info.node_manager_port() << " object manager at "
-                 << node_info.node_manager_address() << ":"
-                 << node_info.object_manager_port();
-  ;
-  RAY_RETURN_NOT_OK(gcs_client_->client_table().Connect(node_info));
+void Raylet::Stop() {
+  RAY_CHECK_OK(gcs_client_->Nodes().UnregisterSelf());
+  acceptor_.close();
+}
```

Comment:
Do Start() and  Stop() have to be public methods? Could they be put into constructor and destructor respectively?

Reply:
1. Method `Start` and `Stop` includes network operations, it's too heavy to put into constructor or destructor.
2. It's convenient to manage the execution order.